pkg/ccl/crosscluster/producer/range_stats.go

```diff
@@ -0,0 +1,112 @@
+// Copyright 2024 The Cockroach Authors.
+//
+// Use of this software is governed by the CockroachDB Software License
+// included in the /LICENSE file.
+
+package producer
+
+import (
+	"context"
+	"sync/atomic"
+	"time"
+
+	"github.com/cockroachdb/cockroach/pkg/repstream/streampb"
+	"github.com/cockroachdb/cockroach/pkg/roachpb"
+	"github.com/cockroachdb/cockroach/pkg/util/ctxgroup"
+	"github.com/cockroachdb/cockroach/pkg/util/hlc"
+	"github.com/cockroachdb/cockroach/pkg/util/log"
+	"github.com/cockroachdb/cockroach/pkg/util/rangedesc"
+	"github.com/cockroachdb/cockroach/pkg/util/span"
+)
+
+// rangeStatsPoller manages a goroutine that polls the total number of ranges
+// and their scanning status. Close must be called to avoid leaking the
+// goroutine.
+type rangeStatsPoller struct {
+	cancel func()
+	g      ctxgroup.Group
+	stats  atomic.Pointer[streampb.StreamEvent_RangeStats]
+}
+
+func startStatsPoller(
+	ctx context.Context,
+	interval time.Duration,
+	spans []roachpb.Span,
+	frontier span.Frontier,
+	ranges rangedesc.IteratorFactory,
+) *rangeStatsPoller {
+	ctx, cancel := context.WithCancel(ctx)
+	poller := &rangeStatsPoller{
+		cancel: cancel,
+		g:      ctxgroup.WithContext(ctx),
+	}
+	poller.g.GoCtx(func(ctx context.Context) error {
+		timer := time.NewTimer(interval)
+		defer timer.Stop()
+		for {
+			stats, err := computeRangeStats(ctx, spans, frontier, ranges)
+			if err != nil {
+				log.Warningf(ctx, "event stream unable to calculate range stats: %v", err)
+			} else {
+				poller.stats.Store(&stats)
+			}
+
+			log.VEventf(ctx, 1, "publishing range stats: %+v", stats)
+
+			select {
+			case <-ctx.Done():
+				return ctx.Err()
+			case <-timer.C:
+				//continue
+			}
+		}
+	})
+	return poller
+}
+
+// Close cancels the internal context and waits for the goroutine to exit.
+func (r *rangeStatsPoller) Close() {
+	r.cancel()
+	_ = r.g.Wait()
```

Comment:
In the current implementation, the only error this could be is cancelled, since it already logs errors as they occur. The code is really just using this as a wait group.